packages/core/src/lib/code-snippet.svelte

```diff
@@ -0,0 +1,220 @@
+</script>
+
+<script lang="ts">
+import { createEventDispatcher, onMount } from 'svelte';
+import cx from 'classnames';
+
+import IconButton from './button/icon-button.svelte';
+import { useTimeout } from './use-timeout';
+import type { IconName } from './icon/icons';
+
+/**
+ * The language to use for syntax highlighting. Must be a language supported by
+ * prism: https://prismjs.com/#supported-languages
+ */
+export let language: string;
+
+/**
+ * The code to render in the code-snippet.
+ */
+export let code: string;
+
+/**
+ * The theme to use for syntax highlighting. Must be a theme provided by
+ * prism-themes: https://github.com/PrismJS/prism-themes#available-themes
+ *
+ * Currently only the `vs` and `vsc-dark-plus` themes are supported.
+ */
+export let theme: CodeSnippetTheme = 'vs';
+
+/**
+ * Whether or not to show a copy-to-clipboard button.
+ */
+export let showCopyButton = true;
+
+/**
+ * Some prism languages have dependencies. For example, C++ requires C. If the
+ * passed `language` has dependencies, they must be included here to be loaded.
+ *
+ * See: https://prismjs.com/plugins/autoloader/
+ */
+export let dependencies: string[] = [];
+
+/**
+ * We use the prism autoloader to handle loading in language grammar files. The
+ * default path for the grammar files is a CDN link so we don't have to include
+ * grammar files in our bundle. If you prefer to point to your own grammars and
+ * bypass the CDN, define the path to those files.
+ *
+ * See: https://prismjs.com/plugins/autoloader/
+ */
+export let grammarsPath = `https://cdnjs.cloudflare.com/ajax/libs/prism/${PRISM_VERSION}/components/`;
+
+/** Additional CSS classes to pass to the aside. */
+let extraClasses: cx.Argument = '';
+export { extraClasses as cx };
+
+const { set: setCopyTimeout } = useTimeout();
+
+let element: HTMLElement | undefined;
+
+$: copyState = 'copy' as CodeSnippetCopyState;
+
+const dispatch = createEventDispatcher<{
+  copy: { succeeded: boolean; message: string };
+}>();
+
+const copyToClipboard = async () => {
+  try {
+    await navigator.clipboard.writeText(code);
+    copyState = 'copied';
+
+    dispatch('copy', {
+      succeeded: true,
+      message: 'Successfully copied snippet to the clipboard',
+    });
+  } catch {
+    copyState = 'failed';
+
+    dispatch('copy', {
+      succeeded: false,
+      message: 'Failed to copy snippet to the clipboard',
+    });
+  }
+
+  setCopyTimeout(() => {
+    copyState = 'copy';
+  }, 2000);
+};
+
+const highlight = () => {
+  if (element) {
+    /**
+     * We need to reset the inner HTML of the element to the raw value of
+     * `code` so it can rescanned for highlighting from the raw.
+     */
+    element.innerHTML = code;
+    highlightElement(element);
+  }
+};
+
+$: if (code) {
+  highlight();
+}
+
+onMount(async () => {
+  // This throws a warning but it is a lie: https://github.com/vitejs/vite/issues/12001
+  const glob = import.meta.glob('/node_modules/prism-themes/themes/*.css');
```

Comment:
Might want to limit this glob to the files we actually use

```suggestion
  const glob = import.meta.glob('/node_modules/prism-themes/themes/prism-{vs,vsc-dark-plus}.css');
```

Reply:
Ah true, though it does raise the question of if we want to support more themes. Might play around with that and see if we want to open that up or not.

Reply:
Oh so I am glad we tested this, fun fact if you change the theme on one code snippet you change the theme on _all_ of them with this new setup. So ... that may not be great. I guess we need to figure out how to scope those CSS files to the component.

Reply:
I looked through the code base and we never use anything but the `vs` theme, and I just checked the `vsc-dark-plus` theme does not work well with JSON and the background color we are applying, number values are almost unreadable:

<img width="753" alt="Screenshot 2023-10-24 at 5 50 22 PM" src="https://github.com/viamrobotics/prime/assets/1930371/4a6365a4-b57c-41f1-b6ab-05e5b1f4bf9c">

I added JSON later and didn't catch that, all the other languages looks ok with the snippets I provided but they don't test the syntax exhaustively. I may come up with better snippets that illustrate the full syntax of each language as best as possible to help catch these kind of issues.

That said, I think for now the best argument is to just keep things simple and not include the `theme` prop at all. Thoughts @mcous @micheal-parks @mrloureed ?